drivers: video: add format utilities to improve usability

josuah opened the pull request:
Introduce several enhancements for dealing with the video API:

- Documentation of formats inspired from [this LWN article](https://lwn.net/Articles/218798/) **[EDIT: this representation is convenient, but note that the bit packing of RGB565, RGB555, RGB565X, RGB555X are wrong in it]**.

note: this image is wrong for RGB565 format!
![scrot_20241006_203817_655x344](https://github.com/user-attachments/assets/99c005b4-1213-4137-8761-1ff282e57f44)

But in a text-based format to be easy to maintain and colorblindness-friendly, and easy to `grep` through[EDIT: fixed version below]:

```
* | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | Bbbbbbbb | Gggggggg | ...
* | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | Gggggggg | Rrrrrrrr | ...
*
* | gggBbbbb | RrrrrGgg | ...
*
* | Xxxxxxxx Rrrrrrrr Gggggggg Bbbbbbbb | ...
```

- Add a printf() formatter like stdint.h's PRIu32:  `printf("video format: " PRIvfmt, PRIvfmt_arg(&fmt));`
  This is present [elsewhere on Zephyr](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/dsp/print_format.h#L30)

- Convert [`video_fourcc()`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/videodev2.h#n81) to `VIDEO_FOURCC()` and make it visible to the documentation, now that it is [used outside of the header](https://github.com/zephyrproject-rtos/zephyr/pull/77879) itself.

- Add a `VIDEO_FOURCC_FROM_STR()` to easily convert a string from i.e. [Kconfig](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/video/capture/Kconfig#L20) or the devicetree.

Looking forward to your feedback and advises.

josuah committed:
drivers: video: improve video formats documentation

Introduce a new text documentation of the video formats, describing
every bit, gives the position of the most significant bit, outline how
it is cut in bytes, and show the pixel segmentation.

Extra care is taken for the RGB565 which requies paying attention to the
bit endianess as well as the byte endianess.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: Make VIDEO_FOURCC() a documented API

Rename video_fourcc() to VIDEO_FOURCC(), differing from the Linux
implementation, but more compliant with the coding style.

Also introduce a VIDEO_FOURCC_FROM_STR() to generate a FOURCC format
code out of a 4-characters string.

Signed-off-by: Josuah Demangeon <me@josuah.net>

loicpoulain commented on include/zephyr/drivers/video.h:
```diff
@@ -33,6 +34,27 @@
+/**
+ * @brief Print formatter to use in printf() style format strings.
+ *
+ * This prints a video format in the form of "{width}x{height} {fourcc}" format
```
From macro below, should be "{fourcc} {width}x{height}" ?

josuah replied:
You are right, I changed it after seeing how `v4l2-ctl` and the examples show the formats but forgot to update thedoc. I will adjust it along with other doc fixups.
Thank you for pointing it!

josuah commented:
force-push: CI fixes
force-push: rebase

josuah commented:
force-push: use `_BPP` instead of `_BITS_PER_PIXEL` following this new function naming:
https://github.com/zephyrproject-rtos/zephyr/pull/76475/files#diff-70c4de6949b97c13ac893b4bfb16fbcc433593fbfa96ec8947bb5cb83c4d04bcR835-R852

josuah commented:
force-push: split `<zephyr/drivers/video-formats.h>` into `<zephyr/drivers/video/formats.h>`, like it is done for all other Zephyr driver includes:  [`include/zephyr/drivers`](https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/drivers)

josuah committed:
drivers: video: introduce video_bits_per_pixel(pixelformat) helper

This was present in the form of video_pix_fmt_bpp() inside ST and NXP
drivers, and was returning the number of bytes, which does not allow
support for 10-bits, 4-bits or non-byte aligned video formats.
The helper leverages the VIDEO_PIX_FMT_*_BITS macros.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- split formats out of `<zephyr/drivers/video.h>` to allow it being included in non-C files, as well as allow the list to grow without cluttering `video.h`, now containing no more "database-like" definitions.
- `_BPP` is replaced by `_BITS` as `_BPP` could mean Bytes Per Pixel or Bits Per Pixel
- Introduce a `video_bits_per_pixel(pixfmt)` using the `_BITS` marcros.
- Fix copyrights (presence only to files changed >50%)

This also propagate the changes to the various files involved.
I can do another variant where the changes are only adding the extra macros/functions, and letting each vendor import the changes progressively.

josuah commented:
force-push:

- Do not split the formats to a separate header for now (only ~150 lines), also to avoid having too many things to review at once. See https://github.com/zephyrproject-rtos/zephyr/pull/78802#pullrequestreview-2385126846 (but glad to reopen a separate PR that takes more time to discuss this!)

josuah commented:
force-push:

- Do not split the formats to a separate header for now (only ~150 lines), also to avoid having too many things to review at once. See https://github.com/zephyrproject-rtos/zephyr/pull/78802#pullrequestreview-2385126846 (but glad to reopen a separate PR that takes more time to discuss this!)

ngphibang commented:
Since https://github.com/zephyrproject-rtos/zephyr/pull/76475 is merged, you will need some rebase. Overall, LGTM.

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -32,6 +32,30 @@
+ * LOG_DBG("Setting format to " PRIvfmt " at %u FPS", PRIvfmt_arg(fmt), fps);
+ * @endcode
+ */
+#define PRIvfmt "%c%c%c%c %ux%u"
```
Good point, the idea was from [`PRIq`](https://github.com/zephyrproject-rtos/zephyr/blob/31fea97e05fd524b089c70281356d07df7788349/include/zephyr/dsp/print_format.h#L30).

I am removing this commit.

josuah commented:
force-push:

- Removing PRIvfmt [as discussed](https://github.com/zephyrproject-rtos/zephyr/pull/79476#discussion_r1816648611)
- Rebase on top of `upstream/main` after https://github.com/zephyrproject-rtos/zephyr/pull/76475 got merged
- Replace new invocation of `video_pix_fmt_bpp()` by `video_bits_per_pixel()`
- Fix line added in commit 1 and  removed in commit 2

About the naming of `video_bits_per_pixel()`, I would have be glad to keep `video_pix_fmt_bpp()`, but I feared that keeping the same function name and changing the meaning (byte -> bit) could cause issue. Glad to use some other name!

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -753,104 +753,184 @@ struct video_buffer *video_buffer_alloc(size_t size);
+ * @endverbatim
+ */
+#define VIDEO_PIX_FMT_XRGB32      VIDEO_FOURCC('B', 'X', '2', '4')
+#define VIDEO_PIX_FMT_XRGB32_BITS 32
```
I understand that you want to help developers who don't need to know each pixel format is of how many bits. But I don't think this is a good idead to define each macro for each format bits. There **will** be many formats and several of them will have the same number of bits, it seems we use too many macros to repeats some numbers (8, 10, 32, etc.).

Another drawback is developer needs to remember the macro name when they can just type, for example 32.

josuah replied:
The logical grouping chosen by `video.h` does not match the grouping in the switch (grouped by number of bits).

It was mostly for facilitating maintenance and avoid mistakes to have each format documented at only one place.
The `switch (pixfmt) { case ...: }` implementation used here was generated to avoid mistakes, and the compiler would likely deduplicate the similar entries.

This does absolutely not matter though, and I am removing the extra macros. :)

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -753,104 +753,184 @@ struct video_buffer *video_buffer_alloc(size_t size);
 
 /**
- * @brief Get number of bytes per pixel of a pixel format
+ * @brief Get number of bit per pixel of a pixel format
```
nits: number of bits

josuah replied:
`variable umber` typo also spotted below...
Thank you for the proofreading!

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -753,104 +753,184 @@ struct video_buffer *video_buffer_alloc(size_t size);
+/**
+ * 5-bit blue followed by 6-bit green followed by 5-bit red, in little-endian over two bytes.
+ * @verbatim
+ * | gggRrrrr | BbbbbGgg | ...
```
I documented the BGR565-LE instead of RGB565-LE, but let's resolve the RGB565 discussion first...
https://github.com/zephyrproject-rtos/zephyr/pull/79996

josuah replied:
Nevermind, this is the RGB565, even if R is on the **right** in the MSB first representation, it is at `BIT(0)`, and seems to be what is expected.

ngphibang replied:
I think this is BGR565-LE.

ngphibang replied:
Endianess is about the byte address order (a chunk of 8 bits), not about the bit order (MSB or LSB first) and neither a chunk of 4 bits.

To get it simple, I think we can begin with BE first. An RGB565 in BE is represented as two bytes with the most significant **byte** is stored at the first **byte**

- RRRRRGGG(1st byte)  GGGBBBBB(2nd byte)

At this point, I think we will all agree with this. Then, LiE is basically reversing the **byte** order for a **multi-byte value**, lower significant bytes get lower address. So it becomes:

- GGGBBBBB(1st byte)  RRRRRGGG(2nd byte)

josuah replied:
What I documented here follows Libav and the LWN article, from where I had the representation:

If representing the format in MSBit first, B is on the left.
```
big-endian, MSBit first:
15 14 13 12 11 10 9  8  7  6  5  4  3  2  1  0
B  b  b  b  b  G  g  g  g  g  g  R  r  r  r  r 

little-endian, MSBit first:
7  6  5  4  3  2  1  0  15 14 13 12 11 10 9  8
g  g  g  R  r  r  r  r  B  b  b  b  b  G  g  g
```

![scrot_20241026_115436_453x25](https://github.com/user-attachments/assets/2a74bbcd-4087-4b6b-bbd2-a56399e33697)
![scrot_20241026_115322_450x30](https://github.com/user-attachments/assets/0ccfa15a-fff2-416a-baa1-dc129b2f1d12)


If representing the format in LSBit first, R is on the left. This representation is confusing when mixing bit-endianness and byte endianness.
```
big-endian, LSBit first:
0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15
r  r  r  r  R  g  g  g  g  g  G  b  b  b  b  B

little-endian, LSBit first:
8  9 10 11 12 13 14 15 0  1  2  3  4  5  6  7
g  g  G  b  b  b  b  B  r  r  r  r  R  g  g  g
```

> At this point, I think we all agree with this.

I am trying to understand what the major implementations did agree... and they *seems* to differ? I am really not sure as they are around since a lot of time. ^_^' Can we settle that the hardware is right to break the tie?

Libavformat *tests source* and this LWN article represent RGB565 with red on the least significant bit:
- https://lwn.net/Articles/218798/
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavfilter/vsrc_testsrc.c#L403

Gstreamer, Intel libvpl, and Libavformat *codecs* represent RGB565 with red on the most significant bit:
- https://github.com/GStreamer/gstreamer/blob/fe61bc3cee1bc7b99b550005f77d8ece3649b4ec/subprojects/gst-plugins-base/gst-libs/gst/video/video-orc.orc#L862-L864
- https://github.com/GStreamer/gstreamer/blob/fe61bc3cee1bc7b99b550005f77d8ece3649b4ec/subprojects/gst-plugins-good/ext/libcaca/gstcacatv.c#L158-L163
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavcodec/bmp.c#L163
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavcodec/xwddec.c#L180
- https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavcodec/dds.c#L369
- The hardware seems to follow what you say.

I think red on MSBit wins! Glad I raised this point.
A test pattern generator looks "ok" (stripes of color) even with the wrong color order, and the LWN article is just text, not code that was checked against hardware. Some deeply anchored confusion in tech it seems!
Good to have this ambiguity tamed once for all, thank you for your patience.

Committing the fix.

ngphibang replied:
Yes, I believe strongly that the LWN article is WRONG. Personally, If I need a reference, I am usually based on the kernel document. The kernel document of the latest Linux release describes the same layout about the RGB565-LE as I pointed in above comment:

https://www.kernel.org/doc/html/v6.11/userspace-api/media/v4l/pixfmt-rgb.html

josuah commented:
force-push:

- Remove the `_BITS` macros [as discussed here](https://github.com/zephyrproject-rtos/zephyr/pull/79476#discussion_r1816808541)
- Typo fixes [as discussed here](https://github.com/zephyrproject-rtos/zephyr/pull/79476#discussion_r1816810278)
- Fix an inversion in a bayer format documentation
- Add extra description of the RGB565 with the bit position to avoid ambiguity [as discussed here](https://github.com/zephyrproject-rtos/zephyr/pull/79996) (glad to make further edits if this is not fully resolved!)
- Add the RGB565X format for the sole purpose of having the big-endian version at sight on the documentation.
- Other typo fixes.

josuah commented:
force-push:
- Doc fix, I did not spot it from running `make -C doc doxygen`!

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -753,104 +753,185 @@ struct video_buffer *video_buffer_alloc(size_t size);
-#define VIDEO_PIX_FMT_RGB565 video_fourcc('R', 'G', 'B', 'P') /* 16  RGB-5-6-5 */
+/**
+ * 5 red bits [4:0], 6 green bits [10:5], 5 blue bits [15:11].
+ * In big endian format over two bytes.
```
Why this format is described in BE while the RGB565 is described in LE, why we don't describe both in either LE or BE ?

josuah replied:
Linux's [`V4L2_PIX_FMT_RGB565X`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/videodev2.h#n564) is the big endian version of [`V4L2_PIX_FMT_RGB565`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/videodev2.h#n560).

This text description documents the bits, then says this should be packed as little-endian, like in the text bit diagram.

This matches the C operations done to decode the format:
```c
uint16_t rgb = sys_be16_to_cpu(rgb565x); // Or sys_le16_to_cpu() for little-endian RGB565
uint8_t r = (rgb >> 11) << 3;
uint8_t g = (rgb >> 5) << 2;
uint8_t b = (rgb >> 0) << 3;
```

Before I figured out the reference material I used (LWN) was wrong, I documented the bits to be "red first", but assigned to the wrong bits to it ([4:0] instead of [15:11]). I adjusted this in the latest commit.

Do you mean you would rather have this description format for i.e. Little Endian?

```
 * bits [2:0] of green, bits [4:0] of blue, bits [4:0] of red, bits [5:3] of green.
```

Do you mean we should opt for RGB565LE and RGB565BE [like done in libav](https://github.com/libav/libav/blob/c4642788e83b0858bca449f9b6e71ddb015dfa5d/libavutil/pixfmt.h#L98-L106) (for instance)? This is different from Linux source but less ambiguous.

ngphibang replied:
Here I just mean the consistency. If we documented RGB565 in LE, so we should also talk about other formats in LE too, or vice versa. But this is just a small detail.

josuah replied:
It looks like insofar, `RGB565` (BE/LE) is the only format where a field spans across multiple bytes.
Other formats are all described by byte access with no big-endian/little-endian variant AFAIU:

- Bayer formats: 8-bit per pixel: no byte endianess issue
- `XRGB32` (and `XYUV` and `YUYV`): There is a `BGRX32`, just like there is a `BGR565`, but no `XRGB32-LE` (or `XRGB32X`):
  This is 8-bit per channel, so different endianess == different channel order

I avoided to add more in case it was considered as unnecessary.
Let me know if I should add i.e. "all little-endian counterparts" or "all of Linux" or "all the most common" etc. :)

I added a comment mentioning that a 16-bit integer had to be split, which does not happen for the other formats.

ngphibang replied:
It's ok. Thanks.

ngphibang commented on include/zephyr/drivers/video.h:
```diff
@@ -753,104 +753,191 @@ struct video_buffer *video_buffer_alloc(size_t size);
+ * @retval bit size of one pixel for this format
  */
-static inline unsigned int video_pix_fmt_bpp(uint32_t pixfmt)
+static inline unsigned int video_bits_per_pixel(uint32_t pixfmt)
```
This one is not an API. It's just an utility function to be used by drivers only. Do we also need migration guide for this ?

ngphibang replied:
Perhaps I misunderstood ... It seems functions in a .h are considered as APIs.

kartben replied:
Anything that's part of a public header and not explicitly marked as internal is to be considered de-facto API :) (see e.g. https://docs.zephyrproject.org/latest/doxygen/html/group__video__interface.html#gabc1c6fd4e13480b269cac9f224ee1c5b)
You may want to review the current API and look for those utilities you folks would like to mark as internal (but that would still require a heads-up to users in the form of a migration guide so that they are aware they might want to move away from using the now internal API).
FWIW it looks to me as if it would be fine to actually keep this as a public API though, it's a utility that might be useful for  any user, not just in-tree drivers..?

ngphibang replied:
yes, it should not be internal and can be used by any user. Thanks

kartben commented:
@josuah will you be coming back to this?

josuah commented:
Yes absolutely. As soon as the "must finish before holidays" race is over if possible.

josuah committed:
doc: releases: 4.1: announce switch to video_bits_per_pixel()

The previous commit introduces video_bits_per_pixel() and converts
all invocation relying on video_pix_fmt_bpp(), 3rd-party users will
need to do the same on their code, and this is not a 100% drop-in API,
so better document it on the release notes.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
force-push:
- rebased on main
- added a release note and migration guide

cfriedt reviewed (changes requested):
Please use `BITS_PER_BYTE` instead of `/8` or `*8`

josuah commented:
force-push:
- rebase on `main`
- use `/ BITS_PER_BYTE` instead of `/ 8` when relevant (thank you @cfriedt!)

I also applied `BITS_PER_BYTE` [here](https://github.com/zephyrproject-rtos/zephyr/pull/79476/commits/db41ca133dee47eb4558525a3cac4b2f7ad069a9#diff-f7423f09c35fbf23beefd1f706f4e1dbd5477daf2c40fbc1d6b731d206156bd9L77) location hoping that is what was meant.

cfriedt commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -331,6 +331,11 @@ Video
+* The ``video_pix_fmt_bpp()`` function was returning a byte count, this got replaced by
+  ``video_bits_per_pixel()`` which return a bit count. For instance, invocations such as
+  ``pitch = width * video_pix_fmt_bpp(pixfmt)`` needs to be replaced by an equivalent
+  ``pitch = width * video_bits_per_pixel(pixfmt) / BITS_PER_PIXEL``.
```
this should probably be BYTE instead of PIXEL
```suggestion
  ``pitch = width * video_bits_per_pixel(pixfmt) / BITS_PER_BYTE``.
```

josuah replied:
My bad, I tried to slip this between two other tasks. I'll wait to be less busy and try again.

josuah commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -228,6 +228,11 @@ Video
   disambiguating.
 
+* The ``video_pix_fmt_bpp()`` function was returning a byte count, this got replaced by
+  ``video_bits_per_pixel()`` which return a bit count. For instance, invocations such as
+  ``pitch = width * video_pix_fmt_bpp(pixfmt)`` needs to be replaced by an equivalent
```
Thank you for teaching distracted prgrmr more English grammar.

kartben commented:
@josuah can you please rebase? random CI glitch causing a job to be stuck and can't seem to be able to re-trigger it manually. Thanks!

josuah commented:
Yes of course!

kartben reviewed (approved):
> Yes of course!

Thanks! Didn't expect it to dismiss existing approvals as I don't think there were any conflicts but, anyway, refresh +1 :)
